src/engine/Distinct.cpp

```diff
@@ -65,75 +66,117 @@ ProtoResult Distinct::computeResult(bool requestLaziness) {
+  auto begin =
+      std::ranges::find_if(result, [this, &previousRow](const auto& row) {
+        // Without explicit this clang seems to
+        // think the this capture is redundant.
+        return !previousRow.has_value() ||
+               !this->matchesRow(row, previousRow.value());
+      });
+  auto end = result.end();
 
   auto dest = result.begin();
-  if (first == dest) {
+  if (begin == dest) {
     // Optimization to avoid redundant move operations.
-    first = std::ranges::adjacent_find(first, last, matchesRow);
-    dest = first;
-    if (first != last) {
-      ++first;
+    begin = std::ranges::adjacent_find(begin, end,
+                                       [this](const auto& a, const auto& b) {
+                                         // Without explicit this clang seems to
+                                         // think the this capture is redundant.
+                                         return this->matchesRow(a, b);
+                                       });
+    dest = begin;
+    if (begin != end) {
+      ++begin;
     }
-  } else if (first != last) {
-    *dest = std::move(*first);
+  } else if (begin != end) {
+    *dest = std::move(*begin);
   }
 
-  if (first != last) {
-    while (++first != last) {
-      if (!matchesRow(*dest, *first)) {
-        *++dest = std::move(*first);
+  if (begin != end) {
+    while (++begin != end) {
+      if (!matchesRow(*dest, *begin)) {
+        *++dest = std::move(*begin);
+        checkCancellation();
       }
     }
     ++dest;
   }
-  result.erase(dest, last);
+  checkCancellation();
+  result.erase(dest, end);
+  checkCancellation();
 
   LOG(DEBUG) << "Distinct done.\n";
   return std::move(result).toDynamic();
 }
+
+// _____________________________________________________________________________
+template <size_t WIDTH>
+IdTable Distinct::outOfPlaceDistinct(const IdTable& dynInput) const {
+  AD_CONTRACT_CHECK(keepIndices_.size() <= dynInput.numColumns());
+  LOG(DEBUG) << "Distinct on " << dynInput.size() << " elements.\n";
+  auto inputView = dynInput.asStaticView<WIDTH>();
+  IdTableStatic<WIDTH> output{dynInput.numColumns(), allocator()};
+
+  auto begin = inputView.begin();
+  auto end = inputView.end();
+  while (begin < end) {
+    int64_t allowedOffset = std::min(end - begin, CHUNK_SIZE);
+    begin = std::ranges::unique_copy(begin, begin + allowedOffset,
+                                     std::back_inserter(output),
+                                     [this](const auto& a, const auto& b) {
+                                       // Without explicit this clang seems to
+                                       // think the this capture is redundant.
+                                       return this->matchesRow(a, b);
+                                     })
+                .in;
+    checkCancellation();
+    // Skip to next unique value
+    do {
+      allowedOffset = std::min(end - begin, CHUNK_SIZE);
+      auto lastRow = begin[-1];
```

Comment:
Is that always safe,, especially for empty inputs? At least comment it, it looks locally scary.

Reply:
It is probably safe because of the outer `begin < end` so the ùnique_copy advances at least by one, but still comment and have a unit test for this.